kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyBackendHandler.java

```diff
@@ -5,99 +5,170 @@
+        if (ctx.channel().isWritable()) {
+            proxyChannelStateMachine.onServerWritable();
+        }
+        else {
+            proxyChannelStateMachine.onServerUnwritable();
         }
     }
 
     // Called when the outbound channel is active
     @Override
     public void channelActive(ChannelHandlerContext ctx) throws Exception {
         LOGGER.trace("Channel active {}", ctx);
-        this.frontendHandler.onUpstreamChannelActive(ctx);
+        serverCtx = ctx;
+        if (sslContext == null) {
+            proxyChannelStateMachine.onServerActive();
+        }
         super.channelActive(ctx);
     }
 
     @Override
-    public void channelRead(final ChannelHandlerContext ctx, Object msg) {
-        assert blockedOutboundCtx == null;
-        LOGGER.trace("Channel read {}", msg);
-        final Channel inboundChannel = inboundCtx.channel();
-        if (inboundChannel.isWritable()) {
-            inboundChannel.write(msg, inboundCtx.voidPromise());
-            unflushedWrites = true;
+    public void userEventTriggered(
+                                   ChannelHandlerContext ctx,
+                                   Object evt)
+            throws Exception {
+        if (evt instanceof SslHandshakeCompletionEvent sslEvt) {
+            if (sslEvt.isSuccess()) {
+                proxyChannelStateMachine.onServerActive();
+            }
+            else {
+                proxyChannelStateMachine.onServerException(sslEvt.cause());
+            }
         }
-        else {
-            inboundChannel.writeAndFlush(msg, inboundCtx.voidPromise());
-            unflushedWrites = false;
+        super.userEventTriggered(ctx, evt);
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) {
+        proxyChannelStateMachine.onServerInactive();
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+        proxyChannelStateMachine.onServerException(cause);
+    }
+
+    /**
+     * Relieve backpressure on the server connection by turning on auto-read.
+     */
+    public void inboundChannelWritabilityChanged() {
+        if (serverCtx != null) {
+            serverCtx.channel().config().setAutoRead(true);
         }
     }
 
+    @Override
+    public void channelRead(final ChannelHandlerContext ctx, Object msg) {
+        proxyChannelStateMachine.forwardToClient(msg);
+    }
+
     @Override
     public void channelReadComplete(final ChannelHandlerContext ctx) throws Exception {
         super.channelReadComplete(ctx);
-        final Channel inboundChannel = inboundCtx.channel();
-        if (unflushedWrites) {
-            unflushedWrites = false;
-            inboundChannel.flush();
+        proxyChannelStateMachine.serverReadComplete();
+    }
+
+    public void forwardToServer(Object msg) {
+        if (serverCtx == null) {
+            LOGGER.trace("WRITE to server ignored because outbound is not active (msg: {})", msg);
```

Comment:
To self: Why's ignoring the right thing to do?

Reply:
I think its really just a hangover from some the early prototyping, I don't think we can actually get into the sate that this would trigger any more. 

We should probably just remove it.

Reply:
In ecd6777b8513e1435f707976c4c7563f856c77cd I made it throw an exception. I think that is effectively unreachable code but I'd like to run with it there for a while and remove in a future PR or release once we have a bit more confidence that the state machine is correct.